polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -98,3 +116,873 @@ impl<BlockNumber: Default + From<u32>> Default for SchedulerParams<BlockNumber>
+	pub fn corresponds_to(&self, receipt: &CandidateReceiptV2<H>) -> bool
+	where
+		H: PartialEq,
+	{
+		receipt.descriptor == self.descriptor && receipt.commitments_hash == self.commitments.hash()
+	}
+}
+
+impl PartialOrd for CommittedCandidateReceiptV2 {
+	fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Ord for CommittedCandidateReceiptV2 {
+	fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+		self.descriptor
+			.para_id
+			.cmp(&other.descriptor.para_id)
+			.then_with(|| self.commitments.head_data.cmp(&other.commitments.head_data))
+	}
+}
+
+impl<H: Copy> From<CommittedCandidateReceiptV2<H>> for super::v7::CommittedCandidateReceipt<H> {
+	fn from(value: CommittedCandidateReceiptV2<H>) -> Self {
+		Self { descriptor: value.descriptor.into(), commitments: value.commitments }
+	}
+}
+
+impl<H: Copy> From<CandidateReceiptV2<H>> for super::v7::CandidateReceipt<H> {
+	fn from(value: CandidateReceiptV2<H>) -> Self {
+		Self { descriptor: value.descriptor.into(), commitments_hash: value.commitments_hash }
+	}
+}
+
+/// A strictly increasing sequence number, tipically this would be the least significant byte of the
+/// block number.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+pub struct CoreSelector(pub u8);
+
+/// An offset in the relay chain claim queue.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+pub struct ClaimQueueOffset(pub u8);
+
+/// Default claim queue offset
+pub const DEFAULT_CLAIM_QUEUE_OFFSET: ClaimQueueOffset = ClaimQueueOffset(1);
+
+/// Signals that a parachain can send to the relay chain via the UMP queue.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+pub enum UMPSignal {
+	/// A message sent by a parachain to select the core the candidate is commited to.
+	/// Relay chain validators, in particular backers, use the `CoreSelector` and
+	/// `ClaimQueueOffset` to compute the index of the core the candidate has commited to.
+	SelectCore(CoreSelector, ClaimQueueOffset),
+}
+/// Separator between `XCM` and `UMPSignal`.
+pub const UMP_SEPARATOR: Vec<u8> = vec![];
+
+impl CandidateCommitments {
+	/// Returns the core selector and claim queue offset the candidate has commited to, if any.
+	pub fn selected_core(&self) -> Option<(CoreSelector, ClaimQueueOffset)> {
+		// We need at least 2 messages for the separator and core selector
+		if self.upward_messages.len() < 2 {
+			return None
+		}
+
+		let upward_commitments = self
+			.upward_messages
+			.iter()
+			.cloned()
+			.rev()
+			.take_while(|message| message != &UMP_SEPARATOR)
+			.collect::<Vec<_>>();
+
+		// Check for UMP separator
+		if upward_commitments.len() == self.upward_messages.len() || upward_commitments.is_empty() {
+			return None
+		}
+
+		// Use first commitment
+		let Some(message) = upward_commitments.into_iter().rev().next() else { return None };
```

Comment:
we can simplify this and optimise it to not collect and clone, like so:

```suggestion
        let separator_pos = self
			.upward_messages
			.iter()
			.rposition(|message| message == &UMP_SEPARATOR)?;

		let message = self.upward_messages.get(separator_pos + 1)?;
```